core/order.go

```diff
@@ -567,59 +664,299 @@ func (n *OpenBazaarNode) getPriceInSatoshi(price *pb.Listing_Price) (uint64, err
 	return uint64(satoshis), nil
 }
 
-func verifySignaturesOnListing(contract *pb.RicardianContract) error {
-	for n, listing := range contract.VendorListings {
-		guidPubkeyBytes := listing.VendorID.Pubkeys.Guid
-		bitcoinPubkeyBytes := listing.VendorID.Pubkeys.Bitcoin
-		guid := listing.VendorID.Guid
+func verifySignaturesOnOrder(contract *pb.RicardianContract) error {
+	guidPubkeyBytes := contract.BuyerOrder.BuyerID.Pubkeys.Guid
+	bitcoinPubkeyBytes := contract.BuyerOrder.BuyerID.Pubkeys.Bitcoin
+	guid := contract.BuyerOrder.BuyerID.Guid
+	ser, err := proto.Marshal(contract.BuyerOrder)
+	if err != nil {
+		return err
+	}
+	hash := sha256.Sum256(ser)
+	guidPubkey, err := crypto.UnmarshalPublicKey(guidPubkeyBytes)
+	if err != nil {
+		return err
+	}
+	bitcoinPubkey, err := btcec.ParsePubKey(bitcoinPubkeyBytes, btcec.S256())
+	if err != nil {
+		return err
+	}
+	var guidSig []byte
+	var bitcoinSig *btcec.Signature
+	var sig *pb.Signatures
+	sigExists := false
+	for _, s := range contract.Signatures {
+		if s.Section == pb.Signatures_ORDER {
+			sig = s
+			sigExists = true
+		}
+	}
+	if !sigExists {
+		return errors.New("Contract does not contain a signature for the order")
+	}
+	guidSig = sig.Guid
+	bitcoinSig, err = btcec.ParseSignature(sig.Bitcoin, btcec.S256())
+	if err != nil {
+		return err
+	}
+	valid, err := guidPubkey.Verify(ser, guidSig)
+	if err != nil {
+		return err
+	}
+	if !valid {
+		return errors.New("Buyers's guid signature on contact failed to verify")
+	}
+	checkKeyHash, err := guidPubkey.Hash()
+	if err != nil {
+		return err
+	}
+	guidMH, err := mh.FromB58String(guid)
+	if err != nil {
+		return err
+	}
+	if !bytes.Equal(guidMH, checkKeyHash) {
+		return errors.New("Public key in order does not match reported buyer ID")
+	}
+	valid = bitcoinSig.Verify(hash[:], bitcoinPubkey)
+	if !valid {
+		return errors.New("Buyer's bitcoin signature on contact failed to verify")
+	}
+
+	return nil
+}
+
+func (n *OpenBazaarNode) ValidateOrder(contract *pb.RicardianContract) error {
+	listingMap := make(map[string]*pb.Listing)
+
+	// Check order contains all required fields
+	if contract.BuyerOrder.Payment == nil {
+		return errors.New("Order doesn't contain a payment")
+	}
+	if contract.BuyerOrder.BuyerID == nil {
+		return errors.New("Order doesn't contain a buyer ID")
+	}
+	if len(contract.BuyerOrder.Items) == 0 {
+		return errors.New("Order hasn't selected any items")
+	}
+	if len(contract.BuyerOrder.RatingKey) != 33 {
+		return errors.New("Invalid rating key in order")
+	}
+	if contract.BuyerOrder.Timestamp == nil {
+		return errors.New("Order is missing a timestamp")
+	}
+
+	// Validate that the hash of the items in the contract match claimed hash in the order
+	var itemHashes []string
+	for _, item := range contract.BuyerOrder.Items {
+		exists := false
+		for _, hash := range itemHashes {
+			if hash == item.ListingHash {
+				exists = true
```

Comment:
Seems like we could `continue` the outer `for` here. Saves iterations and then we can remove the `if !exists` check.